tensor_genn/algorithms/weight_normalization/data_based.py

```diff
@@ -0,0 +1,48 @@
+import numpy as np 
+import math
+
+import tensorflow as tf 
+import tensorflow.keras.backend as K
+from pygenn import genn_model, genn_wrapper
+
+'''
+References: 
+Peter U. Diehl, Daniel Neil, Jonathan Binas, Matthew Cook, Shih-Chii Liu, and Michael Pfeiffer. 2015. 
+Fast-Classifying, High-Accuracy Spiking Deep Networks Through Weight and Threshold Balancing. IJCNN (2015)
+'''
+
+class DataBased():
+    def __init__(self,data=None, batch_size=100):
+        self.x_train = data
+        self.batch_size = batch_size # Adjust based on available system memory
+
+    def normalize(self,tf_model):
+        # Get parameterized layer indices
+        wts_exist = [layer.get_weights() != [] for layer in tf_model.layers]
+        wts_exist_inds = [i for i, x in enumerate(wts_exist) if x == True]
+
+        # Function to fetch ReLU outputs from intermediate layers
+        get_output_from_layer = K.function([tf_model.layers[0].input],
+                                           [tf_model.layers[i].output for i in wts_exist_inds])
+
+        n_batches = math.ceil(len(self.x_train)/self.batch_size)
```

Comment:
It would be really great to include a longer comment here explaining the algorithm at a higher level i.e. saying that, across all the training data, you find the maximum activation of each layer, multiple this by the weights in the subsequent layer and then divide by whatever :smile: